File: go/chip8/keypad.go
package chip8

// Keypad is the state of the 16-key hexadecimal keypad.
type Keypad struct {
	state uint16
}

// Key reports whether key is pressed; keys outside 0x0 - 0xF are never pressed.
func (k *Keypad) Key(key byte) bool {
	return key < 16 && k.state&(1<<key) != 0
}

func (k *Keypad) Set(key byte, pressed bool) {
	if key >= 16 {
		return
	}
	if pressed {
		k.state |= 1 << key
	} else {
		k.state &^= 1 << key
	}
}

File: rust/src/keypad.rs
/// State of the 16-key hexadecimal keypad.
#[derive(Default)]
pub struct Keypad {
    state: u16,
}

impl Keypad {
    pub fn key(&self, key: u8) -> bool {
        key < 16 && self.state & (1 << key) != 0
    }

    pub fn set(&mut self, key: u8, pressed: bool) {
        if key >= 16 {
            return;
        }
        if pressed {
            self.state |= 1 << key;
        } else {
            self.state &= !(1 << key);
        }
    }
}
